Mathlib/CategoryTheory/Abelian/Projective/Dimension.lean

```diff
@@ -199,3 +205,129 @@ instance (X Y : C) (n : ℕ) [HasProjectiveDimensionLT X n]
     (by assumption) (by assumption)
 
 end CategoryTheory
+
+section ProjectiveDimension
+
+open CategoryTheory
+
+variable {C : Type u} [Category.{v, u} C] [Abelian C]
+
+/-- The projective dimension of object of abelian category. -/
+noncomputable def projectiveDimension (X : C) : WithBot ℕ∞ :=
+  sInf {n : WithBot ℕ∞ | ∀ (i : ℕ), n < i → HasProjectiveDimensionLT X i}
+
+lemma projectiveDimension_eq_of_iso {X Y : C} (e : X ≅ Y) :
+    projectiveDimension X = projectiveDimension Y := by
+  simp only [projectiveDimension]
+  congr! 5
+  exact ⟨fun h ↦ hasProjectiveDimensionLT_of_iso e _,
+    fun h ↦ hasProjectiveDimensionLT_of_iso e.symm _⟩
+
+lemma hasProjectiveDimensionLT_of_projectiveDimension_lt (X : C) (n : ℕ)
+    (h : projectiveDimension X < n) : HasProjectiveDimensionLT X n := by
+  have : projectiveDimension X ∈ _ := csInf_mem (by
+    use ⊤
+    simp)
+  simp only [Set.mem_setOf_eq] at this
+  exact this n h
+
+lemma projectiveDimension_le_iff (X : C) (n : ℕ) : projectiveDimension X ≤ n ↔
+    HasProjectiveDimensionLE X n := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · apply hasProjectiveDimensionLT_of_projectiveDimension_lt X (n + 1)
+    exact lt_of_le_of_lt h (Nat.cast_lt.mpr (lt_add_one n))
+  · apply sInf_le
+    simp only [Set.mem_setOf_eq, Nat.cast_lt]
+    intro i hi
+    exact hasProjectiveDimensionLT_of_ge X (n + 1) i hi
+
+lemma projectiveDimension_ge_iff (X : C) (n : ℕ) : n ≤ projectiveDimension X  ↔
+    ¬ HasProjectiveDimensionLT X n := by
+  refine ⟨fun h ↦ ?_, fun h ↦ ?_⟩
+  · simp only [projectiveDimension, le_sInf_iff, Set.mem_setOf_eq] at h
+    by_contra lt
+    by_cases eq0 : n = 0
+    · have := h ⊥ (fun i _ ↦ (hasProjectiveDimensionLT_of_ge X n i (by simp [eq0])))
+      simp [eq0] at this
+    · have : ∀ (i : ℕ), (n - 1 : ℕ) < (i : WithBot ℕ∞) → HasProjectiveDimensionLT X i := by
+        intro i hi
+        exact hasProjectiveDimensionLT_of_ge X n i (Nat.le_of_pred_lt (Nat.cast_lt.mp hi))
+      have not := Nat.cast_le.mp (h (n - 1 : ℕ) this)
+      omega
+  · simp only [projectiveDimension, le_sInf_iff, Set.mem_setOf_eq]
+    intro m hm
+    by_contra nle
+    exact h (hm _ (lt_of_not_ge nle))
+
+lemma projectiveDimension_eq_bot_iff (X : C) : projectiveDimension X = ⊥ ↔
+    Limits.IsZero X := by
```

Comment:
Some of these lemmas should be proven first for `HasProjectiveDimensionLT X`. When I added `HasProjectiveDimensionLT` initially, there was not yet the comparison between `Ext _ _ 0` and morphisms, so that I did not include the lemma `HasProjectiveDimensionLT X 0 ↔ IsZero X`, but it should be added. Similarly for `Projective`.